File: tests/e2e/pages/SettingsPage.js
import { expect } from "@playwright/test";

import { BasePage } from "./BasePage.js";

/**
 * Page object for the Settings page.
 *
 * Mirrors tests/e2e-python/pages/settings_page.py.
 */
class SettingsPage extends BasePage {
  constructor(page) {
    super(page);

    // Selectors
    this.emailDisplay = '[data-testid="settings-email"]';
    this.errorMessage = '[data-testid="settings-error"]';
    this.successMessage = '[data-testid="settings-success"]';
    this.avatarPreview = '[data-testid="settings-avatar-preview"]';
    this.avatarInput = '[data-testid="settings-avatar-input"]';
    this.clearAvatarButton = '[data-testid="settings-clear-avatar-button"]';
    this.displayNameInput = '[data-testid="settings-display-name-input"]';
    this.bioInput = '[data-testid="settings-bio-input"]';
    this.themeSelect = '[data-testid="settings-theme-select"]';
    this.textDensitySelect = '[data-testid="settings-text-density-select"]';
    this.saveButton = '[data-testid="settings-save-button"]';
    this.deleteAccountButton = '[data-testid="settings-delete-account-button"]';
  }

  /** Navigate to the settings page. */
  async goto() {
    await super.goto("/settings");
  }

  /** Click the save button. */
  async save() {
    await this.page.locator(this.saveButton).click();
  }

  /** Fill in a new display name and save. */
  async updateDisplayName(displayName) {
    await this.page.locator(this.displayNameInput).fill(displayName);
    await this.save();
  }

  /** Fill in a new bio and save. */
  async updateBio(bio) {
    await this.page.locator(this.bioInput).fill(bio);
    await this.save();
  }

  /** Select a theme ("light" or "dark") and save. */
  async changeTheme(theme) {
    await this.page.locator(this.themeSelect).selectOption(theme);
    await this.save();
  }

  /** Select a text density ("normal" or "compact") and save. */
  async changeTextDensity(density) {
    await this.page.locator(this.textDensitySelect).selectOption(density);
    await this.save();
  }

  /** Click the clear-avatar button. */
  async clearAvatar() {
    await this.page.locator(this.clearAvatarButton).click();
  }

  /**
   * Click the delete-account button. The confirm() dialogs this
   * triggers are auto-accepted by setupDialogHandler().
   */
  async deleteAccount() {
    await this.page.locator(this.deleteAccountButton).click();
  }

  /** Assert the "Settings updated successfully!" message appeared. */
  async expectSaveSucceeded() {
    await expect(this.page.locator(this.successMessage)).toBeVisible({
      timeout: 5000,
    });
  }

  /**
   * Assert the given theme is actually applied to the page (the <html>
   * element's data-theme attribute), not just selected in the dropdown.
   */
  async expectThemeApplied(theme) {
    await expect(this.page.locator("html")).toHaveAttribute("data-theme", theme);
  }

  /** Assert the avatar preview reverted to the default image. */
  async expectAvatarIsDefault() {
    await expect(this.page.locator(this.avatarPreview)).toHaveAttribute(
      "src",
      /default-avatar/
    );
  }
}

export { SettingsPage };

File: tests/e2e-python/pages/settings_page.py
"""Page object for the Settings page."""

import re

from playwright.sync_api import Page, expect

from .base_page import BasePage


class SettingsPage(BasePage):
    """Reusable helpers for the account settings page (profile fields,
    theme/density preferences, avatar, and account deletion).
    """

    def __init__(self, page: Page):
        super().__init__(page)

        # Selectors
        self.email_display = '[data-testid="settings-email"]'
        self.error_message = '[data-testid="settings-error"]'
        self.success_message = '[data-testid="settings-success"]'
        self.avatar_preview = '[data-testid="settings-avatar-preview"]'
        self.avatar_input = '[data-testid="settings-avatar-input"]'
        self.clear_avatar_button = '[data-testid="settings-clear-avatar-button"]'
        self.display_name_input = '[data-testid="settings-display-name-input"]'
        self.bio_input = '[data-testid="settings-bio-input"]'
        self.theme_select = '[data-testid="settings-theme-select"]'
        self.text_density_select = '[data-testid="settings-text-density-select"]'
        self.save_button = '[data-testid="settings-save-button"]'
        self.delete_account_button = '[data-testid="settings-delete-account-button"]'

    def goto(self) -> None:
        """Navigate to the settings page."""
        super().goto("/settings")

    def save(self) -> None:
        """Click the save button."""
        self.page.locator(self.save_button).click()

    def update_display_name(self, display_name: str) -> None:
        """Fill in a new display name and save."""
        self.page.locator(self.display_name_input).fill(display_name)
        self.save()

    def update_bio(self, bio: str) -> None:
        """Fill in a new bio and save."""
        self.page.locator(self.bio_input).fill(bio)
        self.save()

    def change_theme(self, theme: str) -> None:
        """Select a theme ("light" or "dark") and save."""
        self.page.locator(self.theme_select).select_option(theme)
        self.save()

    def change_text_density(self, density: str) -> None:
        """Select a text density ("normal" or "compact") and save."""
        self.page.locator(self.text_density_select).select_option(density)
        self.save()

    def clear_avatar(self) -> None:
        """Click the clear-avatar button."""
        self.page.locator(self.clear_avatar_button).click()

    def delete_account(self) -> None:
        """Click the delete-account button. The confirm() dialogs this
        triggers are auto-accepted by the page fixture's dialog handler.
        """
        self.page.locator(self.delete_account_button).click()

    def expect_save_succeeded(self) -> None:
        """Assert the "Settings updated successfully!" message appeared."""
        expect(self.page.locator(self.success_message)).to_be_visible(timeout=5000)

    def expect_theme_applied(self, theme: str) -> None:
        """Assert the given theme is actually applied to the page (the
        <html> element's data-theme attribute), not just selected in the
        dropdown.
        """
        expect(self.page.locator("html")).to_have_attribute("data-theme", theme)

    def expect_avatar_is_default(self) -> None:
        """Assert the avatar preview reverted to the default image."""
        expect(self.page.locator(self.avatar_preview)).to_have_attribute(
            "src", re.compile("default-avatar")
        )
